tests/bluetooth/bsim_bt/bsim_test_l2cap_ecred/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,540 @@
+
+void send_sdu_chan_worker(struct k_work *item)
+{
+	struct channel  *ch =
+		CONTAINER_OF(item, struct channel, work);
+
+	send_sdu(ch->itteration, ch->chan_id, ch->bytes_to_send);
+}
+
+static void send_sdu(int itteration, int chan_idx, int bytes)
+{
+	struct bt_l2cap_chan *chan = &channels[chan_idx].le.chan;
+
+	*((int *)channels[chan_idx].payload) = itteration; /*First 4 bytes in sent payload is itteration count*/
+	if (channels[chan_idx].buf != 0) {
+		LOG_ERR("Buf should have been dealocated by now");
+	}
+	struct net_buf *buf = net_buf_alloc(&tx_data_pool, K_NO_WAIT);
+
+	if (buf == 0) {
+		LOG_ERR("Failed to get buff on ch %i, iteration %i should never happen", chan_idx, chan_idx);
+	}
+	channels[chan_idx].buf = buf;
+	net_buf_reserve(buf, BT_L2CAP_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, channels[chan_idx].payload, bytes);
+
+	LOG_DBG("bt_l2cap_chan_sending ch: %i bytes: %i itteration: %i", chan_idx, bytes, itteration);
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	LOG_DBG("bt_l2cap_chan_send returned: %i", ret);
+
+	if (ret < 0) {
+		LOG_DBG("Error: send failed error: %i", ret);
+		channels[chan_idx].buf = 0;
+		net_buf_unref(buf);
+	}
+}
+
+static void test_peripheral_main(void)
+{
+	int err;
+
+	LOG_DBG("*L2CAP ECRED Peripheral started*");
+	init_workqs();
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+	LOG_DBG("Peripheral Bluetooth initialized.");
+
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+	LOG_DBG("Advertising started.");
+
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+
+	/* Wait a bit to ensure that all LLCP have time to finish */
+	k_sleep(K_MSEC(1000));
+
+	register_l2cap_server();
+
+	connect_num_channels(L2CAP_CHANNELS);
+
+	k_sleep(K_MSEC(500));
+	for (int i = 0; i < SDU_SEND_COUNT; i++) {
+		LOG_DBG("Matv: Iteration %i Sendign on chan0", i);
+		channels[0].itteration = i;
+		channels[0].bytes_to_send = DATA_MTU - 500;
+		k_work_submit_to_queue(&my_work_q0, &channels[0].work);
+		LOG_DBG("Matv: Iteration %i Sendign on chan1", i);
+		channels[1].itteration = i;
+		channels[1].bytes_to_send = DATA_MPS - 2;
+		k_work_submit_to_queue(&my_work_q1, &channels[1].work);
+		k_sleep(K_MSEC(5000));
+	}
+
+	disconnect_all_channels();
+	/* Disconnect */
+	LOG_DBG("Peripheral Disconnecting....");
+	err = bt_conn_disconnect(default_conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	LOG_DBG("Peripheral tried to disconnect");
+	if (err) {
+		FAIL("Disconnection failed (err %d)", err);
+		return;
+	}
+	WAIT_FOR_FLAG_UNSET(is_connected);
+	LOG_DBG("Peripheral Disconnected.");
+	k_sleep(K_MSEC(100));/*give a litle time for central to finish processing of disconnect event*/
+	PASS("L2CAP ECRED Peripheral tests Passed");
+	bs_trace_silent_exit(0);
```

Comment:
Don't think this call is necessary.

Reply:
I don't understand what this call is for really, but if I remove it, the test hangs for minute or so after it is done before the executable returns... since a few other bsim tests call this the same way I will leave this as is... I tried to remove it and fix the hanging issue a few ways, but weren't successful and gave up.